pkg/controller/function/function_controller.go

```diff
@@ -83,62 +99,171 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	buildTemplateName                      = getEnvDefault("BUILD_TEMPLATE", "function-kaniko")
+	_                 reconcile.Reconciler = &ReconcileFunction{}
+
+	// "build and push step"
+	buildAndPushStep = "build-step-build-and-push"
+)
+
+// ReconcileFunction is the controller.Reconciler implementation for Function objects
 // ReconcileFunction reconciles a Function object
 type ReconcileFunction struct {
 	client.Client
 	scheme *runtime.Scheme
 }
 
+func getEnvDefault(envName string, defaultValue string) string {
+	// use default value if environment variable is empty
+	var value string
+	if value = os.Getenv(envName); value == "" {
+		return defaultValue
+	}
+	return value
+}
+
 // Reconcile reads that state of the cluster for a Function object and makes changes based on the state read
 // and what is in the Function.Spec
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups="runtime.kyma-project.io",resources=functions,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups="runtime.kyma-project.io",resources=functions/status,verbs=get;update;patch
 // +kubebuilder:rbac:groups="admissionregistration.k8s.io",resources=mutatingwebhookconfigurations;validatingwebhookconfigurations,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups="serving.knative.dev",resources=services,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups="",resources=serviceaccounts,verbs=get;list
-// +kubebuilder:rbac:groups=";apps;extensions",resources=deployments,verbs=create;get;delete;list;update;patch
+// +kubebuilder:rbac:groups="serving.knative.dev",resources=services;routes;configurations;revisions,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="build.knative.dev",resources=builds;buildtemplates;clusterbuildtemplates;services,verbs=get;list;create;update;delete;patch;watch
+// +kubebuilder:rbac:groups="",resources=serviceaccounts,verbs=get;watch;update;list
+// +kubebuilder:rbac:groups=";apps;extensions",resources=deployments,verbs=create;get;watch;update;delete;list;update;patch
 func (r *ReconcileFunction) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 
-	fnConfigName := "fn-config"
-	fnConfigNamespace := "default"
-	fnConfigNameEnv := os.Getenv("CONTROLLER_CONFIGMAP")
-	fnConfigNamespaceEnv := os.Getenv("CONTROLLER_CONFIGMAP_NS")
+	// Get Function instance
+	fn := &runtimev1alpha1.Function{}
+	err := r.getFunctionInstance(request, fn)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// status of the functon must change to error.
+		r.updateFunctionStatus(fn, runtimev1alpha1.FunctionConditionError)
```

Comment:
for something like this it would make sense to implement a reason field in the status subressource ...

Reply:
True